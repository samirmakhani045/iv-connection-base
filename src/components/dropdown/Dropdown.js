import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';
import Select from 'react-select';

import { useTranslation } from '@studio/hooks/useTranslation';

import {
  DropdownIndicatorsContainer,
  DropdownOption,
  DropdownValueContainer,
} from './dropdown-components';
import { Wrapper, customStyles } from './Dropdown.styles';

const Dropdown = (props) => {
  const {
    disabled,
    multiselect,
    withIcon,
    withError,
    placeholder,
    onChange,
    options,
    ...remainingProps
  } = props;
  const [selectedOption, setSelectedOption] = useState([]);
  const selectRef = useRef();
  const { t } = useTranslation('common.forms.dropdown');

  const handleChange = (selectedOption) => {
    setSelectedOption(selectedOption);
    onChange?.(selectedOption);

    if (multiselect) {
      // timeout required to shift focus from last selected option to input, otherwise click outside does not work
      setTimeout(() => selectRef?.current?.select?.focus(), 100);
    }
  };

  const renderDropdownIndicatorsContainer = (indicatorsContainerProps) => {
    return <DropdownIndicatorsContainer {...indicatorsContainerProps} />;
  };

  const renderDropdownOption = (optionProps) => {
    return <DropdownOption withIcon={withIcon} {...optionProps} />;
  };

  const renderDropdownValueContainer = (valueContainerProps) => {
    return <DropdownValueContainer withIcon={withIcon} {...valueContainerProps} />;
  };

  return (
    <Wrapper
      data-testid="dropdown"
      disabled={disabled}
      isEmpty={!selectedOption?.length && !Object.keys(selectedOption).length}
      withError={withError}
      {...remainingProps}
    >
      <Select
        data-testid="dropdownSelect"
        ref={selectRef}
        isMulti={multiselect}
        isDisabled={disabled}
        isSearchable={false}
        hideSelectedOptions={false}
        closeMenuOnSelect={!multiselect}
        value={selectedOption}
        onChange={handleChange}
        options={options}
        components={{
          IndicatorsContainer: renderDropdownIndicatorsContainer,
          Option: renderDropdownOption,
          ValueContainer: renderDropdownValueContainer,
        }}
        styles={customStyles}
        placeholder={placeholder || t('placeholder')}
        {...remainingProps}
      />
    </Wrapper>
  );
};

Dropdown.propTypes = {
  /** Disables the dropdown */
  disabled: PropTypes.bool,
  /** Allow dropdown to multi select options */
  multiselect: PropTypes.bool,
  /** Render icon on the left side of the dropdown */
  withIcon: PropTypes.bool,
  /** Highlight dropdown border with error color */
  withError: PropTypes.bool,
  /** Text that prompts user to open dropdown */
  placeholder: PropTypes.string,
  /** Callback to be fired when dropdown value changes */
  onChange: PropTypes.func,
  /** Options to populate dropdown */
  options: PropTypes.arrayOf(
    PropTypes.shape({
      /** Value to be sent to server */
      value: PropTypes.string.isRequired,
      /** Text that is visible and readable on the dropdown */
      label: PropTypes.string.isRequired,
      /** Icon to be placed on the left side of the dropdown */
      icon: PropTypes.string,
      /** Text to be placed on the right side of the label */
      pill: PropTypes.string,
    }),
  ),
};

Dropdown.defaultProps = {
  disabled: false,
  multiselect: false,
  withIcon: false,
  withError: false,
  options: [],
};

export default Dropdown;
