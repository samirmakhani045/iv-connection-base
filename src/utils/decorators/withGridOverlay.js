import React from 'react';
import styled from 'styled-components/macro';

import * as colors from '@studio/theme/colors';
import { Grid, Box } from '@studio/components/grid';

const Div = styled.div({
  position: 'relative',
});

const StyledGrid = styled(Grid)({
  height: 'calc(100vh - 2rem)',
  left: '50%',
  pointerEvents: 'none',
  position: 'absolute',
  top: '50vh',
  transform: 'translate(-50%, -50vh)',
  width: '100%',
  zIndex: '1000',
});

const StyledBox = styled(Box)({
  backgroundColor: colors.red,
  height: '100%',
  opacity: '0.1',

  '&:nth-child(odd)': {
    opacity: '0.15',
  },
});

const withGridOverlay = (story, props) => {
  if (!props?.args?.gridOverlay) return <>{story()}</>;

  return (
    <Div>
      {story()}
      <StyledGrid>
        <StyledBox />
        <StyledBox />
        <StyledBox />
        <StyledBox />
        <StyledBox />
        <StyledBox />
        <StyledBox />
        <StyledBox />
        <StyledBox />
        <StyledBox />
        <StyledBox />
        <StyledBox />
        <StyledBox />
        <StyledBox />
        <StyledBox />
        <StyledBox />
      </StyledGrid>
    </Div>
  );
};

export default withGridOverlay;
