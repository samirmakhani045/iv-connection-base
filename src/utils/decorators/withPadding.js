import React from 'react';
import styled from 'styled-components/macro';

const Wrapper = styled.div({
  padding: '3rem',
});

const withPadding = (story) => <Wrapper>{story()}</Wrapper>;

export default withPadding;
