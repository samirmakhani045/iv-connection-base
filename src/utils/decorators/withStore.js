import React from 'react';
import { Provider } from 'react-redux';

import configureStore from '@studio/store/configureStore';

const withStore = (story) => {
  const store = configureStore();

  return <Provider store={store}>{story()}</Provider>;
};

export default withStore;
