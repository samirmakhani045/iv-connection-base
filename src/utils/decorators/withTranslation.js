import React from 'react';
import { Provider } from 'react-redux';
import { I18nextProvider } from 'react-i18next';

import i18n from '@studio/i18n';
import localTranslations from '@studio/translations';
import configureStore from '@studio/store/configureStore';
import { flatten } from '@utils/flatten';

const withTranslation = (story) => {
  const store = configureStore();
  i18n.addResources('en', 'translations', flatten(localTranslations));

  return (
    <I18nextProvider i18n={i18n}>
      <Provider store={store}>{story()}</Provider>
    </I18nextProvider>
  );
};

export default withTranslation;
