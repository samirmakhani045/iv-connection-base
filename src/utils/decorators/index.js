import withGridOverlay from './withGridOverlay';
import withPadding from './withPadding';
import withStore from './withStore';
import withTranslation from './withTranslation';

export { withGridOverlay, withPadding, withStore, withTranslation };
