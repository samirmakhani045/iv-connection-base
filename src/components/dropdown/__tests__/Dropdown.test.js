import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { withTranslation } from '@utils/decorators';
import { options } from '@stories/form/dropdown/data.mock';

import Dropdown from '../Dropdown';

const defaultProps = {
  disabled: false,
  multiselect: false,
  withIcon: true,
  withError: false,
  options,
};

const Test = (props) => withTranslation(() => <Dropdown {...defaultProps} {...props} />);

describe('<Dropdown />', () => {
  it('should render correctly', () => {
    const { container } = render(<Test />);

    expect(screen.queryByTestId('dropdown')).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render a disabled dropdown', () => {
    const { container } = render(<Test disabled />);

    expect(screen.queryByTestId('dropdownIndicatorsContainerDisabled')).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render a dropdown with error', () => {
    const { container } = render(<Test withError />);

    expect(screen.queryByTestId('dropdown')).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render a dropdown with icons', () => {
    render(<Test withIcon />);

    expect(screen.queryByTestId('dropdownValueContainerIcon')).toBeInTheDocument();
  });

  it('should fire onChange if dropdown has callback', async () => {
    const onChange = jest.fn();
    render(<Test onChange={onChange} />);

    const dropdown = screen.queryByTestId('dropdown');
    expect(dropdown).toBeDefined();
    expect(onChange).toHaveBeenCalledTimes(0);

    fireEvent.keyDown(dropdown.firstChild, { key: 'ArrowDown' });
    await screen.findByText(options[0].label);
    userEvent.click(screen.getByText(options[0].label));

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith(options[0]);
  });

  it('should select multiple options', async () => {
    const onChange = jest.fn();
    render(<Test onChange={onChange} />);

    const dropdown = screen.queryByTestId('dropdown');
    expect(dropdown).toBeDefined();
    expect(onChange).toHaveBeenCalledTimes(0);

    fireEvent.keyDown(dropdown.firstChild, { key: 'ArrowDown' });
    await screen.findByText(options[0].label);
    userEvent.click(screen.getByText(options[0].label));

    fireEvent.keyDown(dropdown.firstChild, { key: 'ArrowDown' });
    await screen.findByText(options[1].label);
    userEvent.click(screen.getByText(options[1].label));

    expect(onChange).toHaveBeenCalledTimes(2);
  });
});
