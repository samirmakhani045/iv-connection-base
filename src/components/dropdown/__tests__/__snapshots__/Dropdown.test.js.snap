// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> should render a disabled dropdown 1`] = `
<div
  class="Dropdownstyles__Wrapper-evishe-0 hGkWrh"
  data-testid="dropdown"
  disabled=""
>
  <div
    class=" css-14jk2my-container"
  >
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      class="css-1f43avz-a11yText-A11yText"
    />
    <div
      class=" css-1wzx6za-Control"
    >
      <div
        class=" css-wx473i-ValueContainer"
      >
        <div
          class=" css-1xk0z9w-SingleValue"
        >
          <div
            class="DropdownComponentsstyles__IconWrapper-sc-1uiyezv-2 gbFrsL"
            data-testid="dropdownValueContainerIcon"
          >
            <div
              class="DropdownComponentsstyles__NoIcon-sc-1uiyezv-3 jbuDVa"
            />
          </div>
          <div
            class=" css-1onk2zo-Placeholder"
          >
            Select
          </div>
          <input
            aria-autocomplete="list"
            class="css-wmatm6-dummyInput-DummyInput"
            disabled=""
            id="react-select-3-input"
            readonly=""
            tabindex="0"
            value=""
          />
        </div>
      </div>
      <div
        class=" css-ar9wxe-IndicatorsContainer"
      >
        <svg
          data-testid="dropdownIndicatorsContainerDisabled"
          fill="#798B9B"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M6 20H18V10H6V20ZM12 13C13.1 13 14 13.9 14 15C14 16.1 13.1 17 12 17C10.9 17 10 16.1 10 15C10 13.9 10.9 13 12 13Z"
            opacity="0.3"
          />
          <path
            d="M18 8H17V6C17 3.24 14.76 1 12 1C9.24 1 7 3.24 7 6V8H6C4.9 8 4 8.9 4 10V20C4 21.1 4.9 22 6 22H18C19.1 22 20 21.1 20 20V10C20 8.9 19.1 8 18 8ZM9 6C9 4.34 10.34 3 12 3C13.66 3 15 4.34 15 6V8H9V6ZM18 20H6V10H18V20ZM12 17C13.1 17 14 16.1 14 15C14 13.9 13.1 13 12 13C10.9 13 10 13.9 10 15C10 16.1 10.9 17 12 17Z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Dropdown /> should render a dropdown with error 1`] = `
<div
  class="Dropdownstyles__Wrapper-evishe-0 Zbgxb"
  data-testid="dropdown"
>
  <div
    class=" css-2b097c-container"
  >
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      class="css-1f43avz-a11yText-A11yText"
    />
    <div
      class=" css-1wzx6za-Control"
    >
      <div
        class=" css-3xifye-ValueContainer"
      >
        <div
          class=" css-1xk0z9w-SingleValue"
        >
          <div
            class="DropdownComponentsstyles__IconWrapper-sc-1uiyezv-2 gbFrsL"
            data-testid="dropdownValueContainerIcon"
          >
            <div
              class="DropdownComponentsstyles__NoIcon-sc-1uiyezv-3 jbuDVa"
            />
          </div>
          <div
            class=" css-1pa6nd8-Placeholder"
          >
            Select
          </div>
          <input
            aria-autocomplete="list"
            class="css-wmatm6-dummyInput-DummyInput"
            id="react-select-4-input"
            readonly=""
            tabindex="0"
            value=""
          />
        </div>
      </div>
      <div
        class=" css-ar9wxe-IndicatorsContainer"
      >
        <svg
          data-testid="dropdownIndicatorsContainerIcon"
          fill="#798B9B"
          height="1em"
          transform="rotate(90)"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M9.29492 16.885L13.8749 12.295L9.29492 7.70498L10.7049 6.29498L16.7049 12.295L10.7049 18.295L9.29492 16.885Z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Dropdown /> should render correctly 1`] = `
<div
  class="Dropdownstyles__Wrapper-evishe-0 dtiazv"
  data-testid="dropdown"
>
  <div
    class=" css-2b097c-container"
  >
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      class="css-1f43avz-a11yText-A11yText"
    />
    <div
      class=" css-1wzx6za-Control"
    >
      <div
        class=" css-3xifye-ValueContainer"
      >
        <div
          class=" css-1xk0z9w-SingleValue"
        >
          <div
            class="DropdownComponentsstyles__IconWrapper-sc-1uiyezv-2 gbFrsL"
            data-testid="dropdownValueContainerIcon"
          >
            <div
              class="DropdownComponentsstyles__NoIcon-sc-1uiyezv-3 jbuDVa"
            />
          </div>
          <div
            class=" css-1pa6nd8-Placeholder"
          >
            Select
          </div>
          <input
            aria-autocomplete="list"
            class="css-wmatm6-dummyInput-DummyInput"
            id="react-select-2-input"
            readonly=""
            tabindex="0"
            value=""
          />
        </div>
      </div>
      <div
        class=" css-ar9wxe-IndicatorsContainer"
      >
        <svg
          data-testid="dropdownIndicatorsContainerIcon"
          fill="#798B9B"
          height="1em"
          transform="rotate(90)"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M9.29492 16.885L13.8749 12.295L9.29492 7.70498L10.7049 6.29498L16.7049 12.295L10.7049 18.295L9.29492 16.885Z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
